packages/hebao_v3/contracts/base/libwallet/MetaTxLib.sol

```diff
@@ -0,0 +1,156 @@
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../lib/AddressUtil.sol";
+import "../../lib/EIP712.sol";
+import "../../lib/ERC20.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/SignatureUtil.sol";
+import "../../thirdparty/BytesUtil.sol";
+import "./WalletData.sol";
+import "./ERC20Lib.sol";
+import "./QuotaLib.sol";
+
+
+/// @title MetaTxLib
+/// @dev A module to support wallet meta-transactions.
+library MetaTxLib
+{
+    using AddressUtil   for address;
+    using BytesUtil     for bytes;
+    using MathUint      for uint;
+    using SignatureUtil for bytes32;
+    using QuotaLib      for Wallet;
+    using ERC20Lib      for Wallet;
+
+    bytes32 public constant META_TX_TYPEHASH = keccak256(
+        "MetaTx(address relayer,address to,uint256 nonce,address gasToken,uint256 gasPrice,uint256 gasLimit,uint256 gasOverhead,bytes data)"
+    );
+
+    event MetaTxExecuted(
+        address relayer,
+        uint    nonce,
+        bool    success,
+        uint    gasUsed
+    );
+
+    struct MetaTx
+    {
+        address to;
+        uint    nonce;
+        address gasToken;
+        uint    gasPrice;
+        uint    gasLimit;
+        uint    gasOverhead;
+        bool    requiresSuccess;
+        bytes   data;
+        bytes   signature;
+    }
+
+    function validateMetaTx(
+        Wallet  storage wallet,
+        bytes32 DOMAIN_SEPARATOR,
+        MetaTx  memory  metaTx
+        )
+        public
+        view
+    {
+        bytes memory encoded = abi.encode(
+            META_TX_TYPEHASH,
+            msg.sender,
+            metaTx.to,
+            metaTx.nonce,
+            metaTx.gasToken,
+            metaTx.gasPrice,
+            metaTx.gasLimit,
+            metaTx.gasOverhead,
+            metaTx.requiresSuccess,
+            keccak256(metaTx.data)
+        );
+        bytes32 metaTxHash = EIP712.hashPacked(DOMAIN_SEPARATOR, encoded);
+        require(metaTxHash.verifySignature(wallet.owner, metaTx.signature), "INVALID_SIGNATURE");
+    }
+
+    function executeMetaTx(
+        Wallet      storage wallet,
+        bytes32             DOMAIN_SEPARATOR,
+        PriceOracle         priceOracle,
+        MetaTx      memory  metaTx
+        )
+        public
+        returns (bool success)
+    {
+        require(!wallet.locked, "WALLET_LOCKED");
+        require(metaTx.to == address(this));
+
+        uint gasLeft = gasleft();
+        require(gasLeft >= (metaTx.gasLimit.mul(64) / 63), "OPERATOR_INSUFFICIENT_GAS");
+
+        // Update the nonce before the call to protect against reentrancy
+        if (metaTx.nonce != 0) {
+            require(isNonceValid(wallet, msg.sender, metaTx.nonce), "INVALID_NONCE");
+            wallet.nonce[msg.sender] = metaTx.nonce;
+        }
+
+        validateMetaTx(
+            wallet,
+            DOMAIN_SEPARATOR,
+            metaTx
+        );
+
+        (success, ) = metaTx.to.call{gas : metaTx.gasLimit}(metaTx.data);
```

Comment:
what if we are calling `executeMetaTx` inside another meta-tx. Should we disallow it?

Reply:
Maybe we should check "msg.sender != address(this)" inside executeMetaTx.

Reply:
Should be fine in theory, but yeah in practice I don't see any benefit so we may as well disallow it just to be safe.